src/Mvc/Mvc.Razor/src/TagHelpers/UrlResolutionTagHelper.cs

```diff
@@ -273,58 +268,35 @@ protected bool TryResolveUrl([StringSyntax(StringSyntaxAttribute.Uri, UriKind.Re
         return true;
     }
 
-    private static int FindRelativeStart(string url)
+    private static bool TryCreateTrimmedString(string input, [NotNullWhen(true)] out string? trimmed)
     {
-        if (url == null || url.Length < 2)
+        trimmed = null;
+        if (input == null)
         {
-            return -1;
+            return false;
         }
 
-        var maxTestLength = url.Length - 2;
-
-        var start = 0;
-        for (; start < url.Length; start++)
+        var url = input.AsSpan();
+        var start = url.IndexOfAnyExcept(ValidAttributeWhitespaceChars);
+        if (start < 0)
         {
-            if (start > maxTestLength)
-            {
-                return -1;
-            }
-
-            if (!IsCharWhitespace(url[start]))
-            {
-                break;
-            }
-        }
-
-        // Before doing more work, ensure that the URL we're looking at is app-relative.
-        if (url[start] != '~' || url[start + 1] != '/')
-        {
-            return -1;
+            return false;
         }
 
-        return start;
-    }
+        // Url without leading whitespace.
+        url = url.Slice(start);
 
-    private static string CreateTrimmedString(string input, int start)
-    {
-        var end = input.Length - 1;
-        for (; end >= start; end--)
+        // Before doing more work, ensure that the URL we're looking at is app-relative.
+        if (url.Length < 2 || !url[..2].SequenceEqual("~/"))
```

Comment:
Does this have some advantage over `StartsWith`?

Reply:
I need to check.
+ I realize there is a `TrimStart()` on `Span which I'd like to measure and compare.

Reply:
Yes, 1st question answered: `StartsWith` is faster. Going to update the PR soon.

Reply:
Second question answered: span's `Trim()` has similar perf as the original code.

Reply:
The other thing I tried is to replace `Substring()` with something like `return input.Length == url.Length ? input : url.ToString()` where url was sliced already. In the my focused benchmarks, this was slightly better for the "happy path" when url.Length == input.Length, but when applied, it had no or a slightly negative impact in the non-happy path.
Hence remained with `Substring()`